### DISCUSSION TITLE
[TypeScript] Copilot is unable to look ahead; produces redundant code

### DISCUSSION THREAD

**Discussion by ffxsam** [community/community#9815] [Copilot Conversations]
This is my biggest pet peeve with Copilot. If I'm in the middle of a block of some sort, it has no ability to look ahead and see what's there, so it'll suggest adding a closing brace/parenthesis when there already is one.

![CleanShot 2022-01-07 at 12 00 45](https://user-images.githubusercontent.com/12532733/148587046-9cd3c5a7-ac96-4e7f-a140-f8fd5c50f53b.png)

**Comment by Urethramancer** [comment]
It's pretty good at lookahead for me, but has recently picked up a bad habit of adding `")` after printf-style formatting functions. That looks like it could be a language-specific variation of the same bug.

**Reply by ffxsam** [reply]
Good point, definitely language-specific. I updated my title to reflect that this is TypeScript.

**Comment by MrLoh** [comment]
Yeah that’s my main complain about Copilot

**Comment by movy** [comment]
Still happening every single time, quite frustrating. I tend to type faster now just not to give CoPilot a chance to autocomplete my line and break it with unnecessary parenthesis and semis :));};

**Comment by albertopfunk** [comment]
I think something is missing from copilot, it does not analyze the file you are working on, the suggestions are almost guesses, and at times it will just duplicate the last suggestion you chose. I think for this to work, copilot needs to be able to analyze projects you work on, look ahead/behind, etc, for context. Once it does that I think it will change the way everyone writes software because you will be able to use copilot to explain any codebase you are not familiar with and do much more!

**Comment by movy** [comment]
How is it not being addressed? Extra semicolons in JS code is another annoyance. Isn't this AI smart enough to figure out whether a user likes semis or despises them?
